securebanking-openbanking-uk-rs-simulator-server/src/main/java/com/forgerock/securebanking/openbanking/uk/rs/api/backoffice/payment/utils/CustomObjectMapper.java

```diff
@@ -0,0 +1,133 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.securebanking.openbanking.uk.rs.api.backoffice.payment.utils;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializerProvider;
+import com.fasterxml.jackson.databind.deser.std.NumberDeserializers;
+import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.ser.std.StdSerializer;
+import org.joda.time.DateTime;
+import org.joda.time.format.DateTimeFormatter;
+import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer;
+import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+import static org.joda.time.format.ISODateTimeFormat.dateTimeNoMillis;
+
+public class CustomObjectMapper {
+    private static CustomObjectMapper customObjectMapper;
+    private static ObjectMapper objectMapper;
+
+    private CustomObjectMapper() {
+        Jackson2ObjectMapperBuilderCustomizer customizer =
+                jacksonObjectMapperBuilder -> {
+                    jacksonObjectMapperBuilder.featuresToEnable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
+                    jacksonObjectMapperBuilder.serializerByType(BigDecimal.class, new BigDecimalSerializer(BigDecimal.class));
+                    jacksonObjectMapperBuilder.deserializerByType(BigDecimal.class, new BigDecimalDeserializer());
+                    jacksonObjectMapperBuilder.serializerByType(DateTime.class, new DateTimeSerializer(DateTime.class));
+                    jacksonObjectMapperBuilder.deserializerByType(DateTime.class, new DateTimeDeserializer());
+                    jacksonObjectMapperBuilder.serializationInclusion(JsonInclude.Include.ALWAYS);
+                };
+
+        Jackson2ObjectMapperBuilder builder = Jackson2ObjectMapperBuilder.json();
+        customizer.customize(builder);
+
+        objectMapper = builder.build();
+        objectMapper.setNodeFactory(JsonNodeFactory.withExactBigDecimals(true));
+    }
+
+    public static CustomObjectMapper getCustomObjectMapper() {
+        if (customObjectMapper == null) {
+            customObjectMapper = new CustomObjectMapper();
+        }
+
+        return customObjectMapper;
+    }
+
+    public static ObjectMapper getObjectMapper() {
+        return objectMapper;
+    }
+
+    public static class BigDecimalSerializer extends StdSerializer<BigDecimal> {
```

Comment:
Can you add some tests for the individual formatters please?

Also it would be nice to have more descriptive names, as these are serializers with logic i.e. Max2DecimalPlacesBigDecimalSerializer and DateTimeNoMillisSerializer

Reply:
I removed my DateTimeSerializer and replaced it with https://github.com/SecureBankingAccessToolkit/securebanking-openbanking-uk/blob/master/securebanking-openbanking-uk-obie-datamodel/src/main/java/uk/org/openbanking/jackson/DateTimeSerializer.java. 

As for the Max2DecimalPlacesBigSerializer I removed the setScale and kept only the transformation to String.

Reply:
Thanks.

What are the requirements for serializing BigDecimals and how does Jackson behave by default?  If it does not behave how we want, then I think there are flags which can be used to customise the behaviour, which would avoid having to write custom serializers.

I've found that we are using BigDecimal to represent the ExchangeRate (see [here](https://github.com/SecureBankingAccessToolkit/securebanking-openbanking-uk/blob/master/securebanking-openbanking-uk-obie-datamodel/src/main/java/uk/org/openbanking/datamodel/payment/OBExchangeRate2.java#L46)), which is defined in the spec [here](https://github.com/SecureBankingAccessToolkit/securebanking-openbanking-uk/blob/master/securebanking-openbanking-uk-obie-datamodel/src/main/resources/specification/3.1.9/payment-initiation-openapi.yaml#L5973-L5975) - in this case the ExchangeRate value in the json document should be a json number and not a string. There's an example consent response [here](https://openbankinguk.github.io/read-write-api-site3/v3.1.9/resources-and-data-models/pisp/international-payment-consents.html#usage-examples)

Reply:
The BigDeciaml formatting will be removed after we finish this unplanned task https://github.com/SecureBankingAccessToolkit/SecureBankingAccessToolkit/issues/483